designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+*Advantages:*
+
+* Extensible to other resource types
+* Supports parallel builds for each resource
+* Aligned with a CloudFormation stack
+
+*Disadvantages:*
+
+* Too many build folders, and hence zip files, to manage.
+* Difficult to share code between all Lambdas.
+
+::
+
+    $PKG_ROOT/build/
+        artifacts.json (not for MVP)
+        MyFunction1/
+             .... <build artifacts>
+
+          MyFunction2/
+             ... <build artifacts>
+
+          MyApiGw/
+            ... <build artifacts>
+
+          MyECRContainer/
+             ... <build artifacts>
+
+
+Future Extensions
+^^^^^^^^^^^^^^^^^
+In the future, we will change the limitation around each folder being named after the resource's LogicalID.
+Instead, we will support an artifacts.json file that will map Lambda function resource's LogicalId to the path to a
+folder that contains built artifacts for this function. This allows us to support custom build systems that use
+different folder layout.
+
+A well-known folder structure also helps “sam local” and “sam package” commands to automatically discover the
+built artifacts for each Lambda function and package it.
+
+Stable Builds
+~~~~~~~~~~~~~
+A build is defined to be stable if the built artifacts changes if and only if the contents or metadata
+(ex: timestamp, ownership) on source files changes. This is an important attribute of a build system. Since SAM CLI
+relies on 3rd party package managers like NPM to do the heavy lifting, we can only provide a "best effort" service here.
+By running ``sam build`` on a build system that creates a new environment from scratch
+(ex: Travis/CircleCI/CodeBuild/Jenkins etc), you can achieve truly stable builds. For more information on why this
+is important, refer to Debian's guide on `reproducible builds <https://reproducible-builds.org/>`_.
+
+
+SAM CLI does the following to produce stable builds:
+
+#. Clean build folder on every run
+#. Include metadata when coping files and folders
+#. Run build actions with minimal information passed from the environment
+
+
+Built-in Build Actions
+~~~~~~~~~~~~~~~~~~~~~~
+Build actions natively supported by SAM CLI follow a standard workflow:
+
+#. Search for a supported dependency manifest file. If a known manifest is not present, we will abort the build.
+#. Setup: Create build folder
+#. Resolve: Install dependencies
+#. Compile: Optionally, compile the code if necessary
+#. Copy Source: Optionally, Copy Lambda function code to the build folder
+
+Setup step is shared among all runtimes. Other steps in the workflow are implemented differently for each runtime.
+
+Javascript using NPM
+^^^^^^^^^^^^^^^^^^^^
+
+Install dependencies specified by ``package.json`` and copy source files
+
+**Manifest Name**: ``package.json``
+
++-------------+--------------------------------------+
+| Action      | Command                              |
++=============+======================================+
+| Resolve     | ``npm install``                      |
```

Comment:
Appreciate the way you're going with this, so we (from Claudia.js) would like to ask if you'd like us to share the code from our `claudia pack` command? 

As just an `npm install` with excluding `node_modules` is not going to be enough. You will have other files, such as `.eslintrc`, still present everywhere.

Reply:
YES YES YES 🔥Thank you so much for offering. This is exciting to get deeper Nodejs build support into the command. The current technical design does not support such integrations. I will rework and send an updated PR.

Broadly, here are some questions I am trying to answer:
- Can we support build actions written natively in other languages (JS/Golang/C#) etc?
- How can we auto-detect certain frameworks and invoke appropriate build actions for that framework because SAM CLI's built-in functionality is never going to be as deep and opinionated as the framework's? (ex: call `claudia pack` for ClaudiaJS apps, or `zappa package` for Zappa apps) 
- How can I let other non-SAM CLI developers write custom build actions?
- How can we build all the custom build actions into SAM CLI so customers don't have to install a plugin?
- If the build command is a collection of 20-30 different build actions for various frameworks, does it make sense to create a separate repository and vend it as a library that SAM CLI and other tools can consume? I know we are never going to ever support all frameworks/opinions applying the 80:20 rule, what would a good 80% look like?

I might be digressing from the original intent of the build command, but seeing the community interest and going through other build/package implementations, I feel like we can all benefit from each other if we collect our efforts.

Thoughts?

Reply:
@sanathkr `claudia pack` will do the proper packaging for any nodejs runtime (dependency installation & rewriting for local deps and removing test resources), so it's not just for projects made with claudia. eg for video puppet, I'm using cloudformation to deploy lambdas with nodejs (not even sam, because of nested templates not working), but I'm using claudia pack to create deployable zips.

Reply:
Ah I didn't know `claudia pack` is generic. Yes, it would be fantastic to be able to adopt that functionality into SAM CLI. 

We have two options:

**Option 1: Include a JS module in SAM CLI's Python source**
To build Node projects, SAM CLI just invoke the JS module `node pack.js`.

Pros: 
- Easy to lift & shift
- Easy to use language specific libraries that can support deeper integrations in future like webpack build or running gulp scripts
- Sets precedence for other runtimes like Java which might need reflexion to create the package
- Easier to get help from JS community who is more familiar with building JS packages.

Cons:
- Vending JS files in Python package
- Might take dependency on certain version of Node. We can't enforce that customers have this version of Node on their system. 
- Might have to webpack all dependencies, minify and vend one file that we just run using `node pack.js`.
- Need to spawn & manage processes
- Could become a tech debt if this approach doesn't scale.

**Option 2: Rewrite the `claudia pack` functionality in Python**

Pros:
- Native implementation. 
- Implementation might create general features that can be used by build actions for other runtimes
- Faster

Cons:
- Takes a lot more work. Can't lift & shift
- Won't get to use language native features.


What do you guys think?

Reply:
I guess we can start with Node.js (option 1), and rewrite in Python if needed later.

For Node.js version, with Claudia.js we try to fully support versions supported by Lambda, but it works in all major versions of Node. Also it’s easy to have multiple versions of Node.js with tools like nvm.

Reply:
I think it's safe to assume that people deploying node.js lambda projects have node.js installed locally. If not, then neither of the two solutions will work, unless you want to fully rewrite NPM in python as well. 
 
Regarding your option 2 and second con, I don't think that is relevant. if you're just thinking of packaging up lambdas, there are no language-native features required for it (apart from loading JSON, which I assume you can do with Python as well). It's just a matter of moving files around, zipping them up, and invoking NPM a few times with the right options. 

How about having an `install` and `package` steps for language-dependent builders? `install` could prepare the builder from python, by effectively creating an local executable in a runtime-specific way (for example, in case of node.js, this would run `npm install` on a subdirectory where you have the packaging script, to pull the right dependencies; in case of Java it could grab the maven shade plugin...). If that step fails, SAM can complain and tell the user that the local environment needs aren't met, so it can't build the project, and potentially direct them to a URL documenting what local tools are required for building that runtime. If that step succeeds, you know that it's possible to build it correctly, and SAM would just need to invoke a system executable produced/configured by the first step.

Reply:
@stojanovic Yes, let's do that. The one thing I underestimated is the level of help existing Javascript libraries will provide to package JS apps. Ex: You could just import a gulpfile as module and run actions without treating it like a CLI. 

This creates an interesting state where SAM CLI's Python package will contain JS libraries, but I think that is fine.

Reply:
@gojko I like your idea of splitting the `install` & `package` phases. I would even split the install phase into `prepare`, `resolve-dependencies`, `copy-source` phases where we need to call NPM to simply `resolve-dependencies` and for everything else we will have default Python implementations.

When we consume Claudia's `pack` functionality, we can work through the details of where to draw the line for each phase. But I think it is worth keeping the `pack` code in Javascript so we can add support for advanced building like running Gulpscripts or Webpack etc without re-implementing all these libraries in Python.

Reply:
I added information about how this will be implemented to the design doc: https://github.com/sanathkr/aws-sam-cli/blob/build-cmd-design/designs/sam_build_cmd.rst#implementation-of-build-actions